src/adm.rs

```diff
@@ -1,39 +1,225 @@
 use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
 use url::Url;
 
-use crate::error::HandlerError;
+use crate::error::{HandlerError, HandlerErrorKind, HandlerResult};
+use crate::server::ServerState;
+use crate::settings::Settings;
+//use crate::server::img_storage;
+
+pub(crate) const DEFAULT: &str = "DEFAULT";
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct AdmTileResponse {
     pub tiles: Vec<AdmTile>,
 }
 
-#[derive(Debug, Deserialize, Serialize)]
+/// Filter criteria for adm Tiles
+#[derive(Default, Clone, Debug)]
+pub struct AdmFilter {
+    /// list of allowed base host strings.
+    pub filter_set: BTreeMap<String, AdmAdvertiserFilterSettings>,
+}
+
+/// The AdmAdvertiserFilterSettings contain the settings for the various
+/// ADM provided partners. These are specified as a JSON formatted hash
+/// that contains the components. A special "DEFAULT" setting provides
+/// information that may be used as a DEFAULT, or commonly appearing set
+/// of data.
+/// See `impl From<Settings>` for details of the structure.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct AdmAdvertiserFilterSettings {
+    pub(crate) advertiser_hosts: Vec<String>,
+    pub(crate) impression_hosts: Vec<String>,
+    pub(crate) position: Option<u8>,
+    pub(crate) include_regions: Vec<String>,
+}
+
+pub(crate) type AdmSettings = HashMap<String, AdmAdvertiserFilterSettings>;
+
+impl AdmFilter {
+    /// Check the impression URL to see if it's valid.
+    ///
+    /// This extends `filter_and_process`
+    fn check_impression(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+    ) -> HandlerResult<()> {
+        let parsed: Url = match tile.impression_url.parse() {
+            Ok(v) => v,
+            Err(e) => {
+                return Err(HandlerErrorKind::UnexpectedImpressionHost(format!(
+                    "Invalid host: {:?} {:?}",
+                    e,
+                    tile.impression_url.to_string()
+                ))
+                .into());
+            }
+        };
+        let host = match parsed.host() {
+            Some(v) => v.to_string(),
+            None => {
+                return Err(HandlerErrorKind::UnexpectedImpressionHost(format!(
+                    "Missing impression host: {:?}",
+                    tile.impression_url
+                ))
+                .into());
+            }
+        };
+        if !filter.impression_hosts.contains(&host) {
+            return Err(HandlerErrorKind::UnexpectedImpressionHost(host).into());
+        }
+        Ok(())
+    }
+
+    /// Filter and process tiles from ADM:
+    ///
+    /// - Returns None for tiles that shouldn't be shown to the client
+    /// - Modifies tiles for output to the client (adding additional fields, etc.)
+    pub fn filter_and_process(&self, tile: AdmTile) -> Option<AdmTile> {
```

Comment:
@jrconlin - Those initial filters look good 👍

We also have a concept of ["approved advertisers"](https://mana.mozilla.org/wiki/pages/viewpage.action?spaceKey=FPS&title=Mozilla+Approved+Advertisers). I think we can also check that here.

Feel free to do this in a follow-up.

Reply:
I was thinking that list would go into the "DEFAULT" section in the configuration. I've found it's best to avoid hard coding values whenever possible.